Add SLO datasources

adrien-f opened the pull request:
Closes #911

Signed-off-by: Adrien FILLON <adrien.fillon@manomano.com>

adrien-f commented:
Hi 👋 

A few notes:

* Quite inspired by what has been done for the monitor Data source
* There's a bit of duplication between the single DS and the multiple DS, let me know if you'd like a refactor on this
* I came across a bit of a refactor here (you moved the tests in its own package/folder), let me know if there was anything else missing
* Exposed fields are basic as discussed in the original issue

Let me know what you think!

nmuesch commented on datadog/data_source_datadog_service_level_objectives.go:
```diff
@@ -0,0 +1,109 @@
+		return err
+	}
+	d.SetId("datadog-service-level-objectives")
+	//d.SetId(fmt.Sprintf("%d", hashcode.String(req.String())))
```
Looks like this may have been leftover?

nmuesch reviewed (commented):
Hey @adrien-f Thanks for the PR, and bearing with us during the recent repo restructure, we're trying to organize things a bit. 🙂 

We're currently running `tfplugindocs` with no flags, but only committing relevant changes as there's a couple final pieces we're working out there. 

I haven't gone fully over the tests yet, but the core code seems good to me. I would like to discuss the single+multiple datasource approach though. It would be nice if we could find a way to get this into a single data-source since the object really is the same its just a difference of how many gets returned.

Perhaps we could have a flag in the config like `multiple: true` (with any name 😉) and differentiate the logic in the READ that way?

nmuesch commented on docs/data-sources/service_level_objective.md:
```diff
@@ -0,0 +1,30 @@
+
+Use this data source to retrieve information about an existing SLO for use in other resources.
+
+
```
To get an HCL config generated here, can you add an example to the examples/data-source folder?

adrien-f commented:
Hi 👋 Sorry for the delay!

I've rebased and added examples.

Regarding the issue of having one more two datasources, I really have no opinion on this 😅 Looking at the providers in the registry like AWS or Azure I see they often have two datasources:

* https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/data-sources/shared_image_versions
* https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/data-sources/shared_image_version
* https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/ec2_instance_type_offering
* https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/ec2_instance_type_offerings

I'm not sure there are any clear guidelines on this, I suppose there's a reason? Maybe you could get some pointers from Hashicorp here?

In any case, thanks again for this 🙂 Let's me know,

nmuesch commented:
Hey @adrien-f Thanks for checking around and providing those links. Given that this pattern looks to be the standard, I think it's OK to proceed with two data sources. 

I plan on taking another pass through the PR, but could you update the branch with `master`? We've added some more CI for linting and docs and it would be good to confirm this is succeeding on those 🙂

nmuesch commented on datadog/data_source_datadog_service_level_objectives.go:
```diff
@@ -0,0 +1,108 @@
+	if err := d.Set("slos", slos); err != nil {
+		return err
+	}
+	d.SetId("datadog-service-level-objectives")
```
I think since the data-source allows filtering, and isn't a "global"/"unique" data-source that returns all SLOs, we'll need to give it a unique ID. Maybe something similar to - https://github.com/DataDog/terraform-provider-datadog/blob/master/datadog/data_source_datadog_security_monitoring_rules.go#L140

nmuesch commented on datadog/tests/data_source_datadog_service_level_objective_test.go:
```diff
@@ -0,0 +1,108 @@
+)
+
+func TestAccDatadogServiceLevelObjectiveDatasource(t *testing.T) {
+	ctx, accProviders := testAccProviders(context.Background(), t)
+	uniq := strings.ReplaceAll(uniqueEntityName(ctx, t), "-", "_")
+	accProvider := testAccProvider(t, accProviders)
+
+	resource.ParallelTest(t, resource.TestCase{
```
Sorry, you caught us in a time of some refactors/improvements to the project 🙂 
Could you update the beginning of these tests to match - https://github.com/DataDog/terraform-provider-datadog/blob/master/datadog/tests/data_source_datadog_dashboard_test.go#L13-L18

nmuesch commented on datadog/tests/data_source_datadog_service_level_objectives_test.go:
```diff
@@ -0,0 +1,108 @@
+)
+
+func TestAccDatadogServiceLevelObjectivesDatasource(t *testing.T) {
+	ctx, accProviders := testAccProviders(context.Background(), t)
+	uniq := strings.ReplaceAll(uniqueEntityName(ctx, t), "-", "_")
+	accProvider := testAccProvider(t, accProviders)
+
+	resource.ParallelTest(t, resource.TestCase{
```
Same here on the test updates 🙇

nmuesch commented on datadog/tests/data_source_datadog_service_level_objectives_test.go:
```diff
@@ -0,0 +1,108 @@
+}
+
+func checkServiceLevelObjectivesDatasourceAttrs(accProvider *schema.Provider, uniq string) resource.TestCheckFunc {
+	return resource.ComposeTestCheckFunc(
+		resource.TestCheckResourceAttrSet("data.datadog_service_level_objectives.foo", "slos.0.id"),
+		resource.TestCheckResourceAttr("data.datadog_service_level_objectives.foo", "slos.0.name", uniq),
+		resource.TestCheckResourceAttr("data.datadog_service_level_objectives.foo", "slos.0.type", "metric"),
+	)
+}
```
I think it'd be helpful here to assert at least 2 SLOs are returned. Would you be able to add a new config for this?

nmuesch commented on datadog/data_source_datadog_service_level_objectives.go:
```diff
@@ -0,0 +1,108 @@
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Description: "Id of Datadog service level objective",
```
```suggestion
							Description: "ID of the Datadog service level objective",
```

nmuesch commented on datadog/data_source_datadog_service_level_objectives.go:
```diff
@@ -0,0 +1,108 @@
+							Computed:    true,
+						},
+						"name": {
+							Description: "Name of Datadog service level objective",
```
```suggestion
							Description: "Name of the Datadog service level objective",
```

nmuesch reviewed (commented):
Overall these new data-sources look good! Thanks for working on this. 
I left some inline notes, but they're mostly docs + test comments.

adrien-f commented:
Heya, apologies for the delay in my latest update. 

I've updated the tests to include extra SLOs, let me know if you're satisfied with this. It also now computes a custom ID, I had to mess with pointers and I've tried to keep it concise enough and I've rebased everything.

Cheers,

zippolyte reviewed (approved):
Hey @adrien-f, sorry for the delay in taking another look at this.
This looks good to me, I've taken the liberty to rebase on master, as we've updated the go client which contained breaking changes. Hope that's fine.

zippolyte commented:
/azp run

zippolyte commented:
/azp run

adrien-f commented:
Awesome! Thanks a lot 🚀 !

apollocreed commented:
Yeah thx a lot much to both of you much appreciated
